public/gametrace.h

```diff
@@ -12,84 +12,439 @@
+	bool IsTouchPersistsEnabled() const						{ return ( m_nCollisionFunctionMask & FCOLLISION_FUNC_ENABLE_TOUCH_PERSISTS ) != 0; }
+
+	bool HasInteractsAsLayer( int nLayerIndex ) const		{ return ( m_nInteractsAs & ( 1ull << nLayerIndex ) ) != 0; }
+	bool HasInteractsWithLayer( int nLayerIndex ) const		{ return ( m_nInteractsWith & ( 1ull << nLayerIndex ) ) != 0; }
+	bool HasInteractsExcludeLayer( int nLayerIndex ) const	{ return ( m_nInteractsExclude & ( 1ull << nLayerIndex ) ) != 0; }
+	void EnableInteractsAsLayer( int nLayer )				{ m_nInteractsAs |= ( 1ull << nLayer ); }
+	void EnableInteractsWithLayer( int nLayer )				{ m_nInteractsWith |= ( 1ull << nLayer ); }
+	void EnableInteractsExcludeLayer( int nLayer )			{ m_nInteractsExclude |= ( 1ull << nLayer ); }
+	void DisableInteractsAsLayer( int nLayer )				{ m_nInteractsAs &= ~( 1ull << nLayer ); }
+	void DisableInteractsWithLayer( int nLayer )			{ m_nInteractsWith &= ~( 1ull << nLayer ); }
+	void DisableInteractsExcludeLayer( int nLayer )			{ m_nInteractsExclude &= ~( 1ull << nLayer ); }
+	
+public:
+	// Which interaction layers do I represent? (e.g. I am a LAYER_INDEX_CONTENTS_PLAYER_CLIP volume)
+	// NOTE: This is analogous to "contents" in source 1  (bit mask of CONTENTS_* or 1<<LAYER_INDEX_*)
+	uint64 m_nInteractsAs;
+	
+	// Which interaction layers do I interact or collide with? (e.g. I collide with LAYER_INDEX_CONTENTS_PASS_BULLETS because I am not a bullet)
+	// NOTE: This is analogous to the "solid mask" or "trace mask" in source 1 (bit mask of CONTENTS_* or 1<<LAYER_INDEX_*)
+	uint64 m_nInteractsWith;
+	
+	// Which interaction layers do I _not_ interact or collide with?  If my exclusion layers match m_nInteractsAs on the other object then no interaction happens.
+	uint64 m_nInteractsExclude;
+	
+	uint32 m_nEntityId; 			// this is the ID of the game entity
+	uint32 m_nOwnerId;  			// this is the ID of the owner of the game entity
+	uint16 m_nHierarchyId; 			// this is an ID for the hierarchy of game entities (used to disable collision among objects in a hierarchy)
+	uint8 m_nCollisionGroup;		// one of the registered collision groups
+	uint8 m_nCollisionFunctionMask;	// set of CollisionFunctionMask_t bits
+};
 
+class CGameTrace
+{
+public:
+	CGameTrace()
+	{
+		Init();
+	}
+
+	void Init()
+	{
+		m_pSurfaceProperties = NULL;
+		m_pEnt = NULL;
+		m_pHitbox = NULL;
+		m_hBody = NULL;
+		m_hShape = NULL;
+		m_nContents = 0;
+		m_BodyTransform.SetToIdentity();
+		m_vHitNormal.Init();
+		m_vHitPoint.Init();
+		m_flHitOffset = 0.0f;
+		m_flFraction = 1.0f;
+		m_nTriangle = -1;
+		m_nHitboxBoneIndex = -1;
+		m_eRayType = RAY_TYPE_LINE;
+		m_bStartInSolid = false;
+		m_bExactHitPoint = false;
+	}
+	
+	// Returns true if there was any kind of impact at all
+	bool DidHit() const 
+	{ 
+		return m_flFraction < 1 || m_bStartInSolid; 
+	}
+	
+public:
+	const CPhysSurfaceProperties *m_pSurfaceProperties;
+	CBaseEntity *m_pEnt;

```

Comment:
This better to be changed to CEntityInstance, since we don't really define CBaseEntity.